python/tvm/topi/cuda/unique.py

```diff
@@ -0,0 +1,384 @@
+    return ib.get()
+
+
+@hybrid.script
+def _calc_num_unique(data):
+    output = output_tensor((1,), "int32")
+    for i in bind("threadIdx.x", 1):
+        output[i] = data[data.shape[0] - 1] + int32(1)
+    return output
+
+
+def _calc_unique_sorted_ir(data, argsorted_indices, inc_scan, unique_elements, indices):
+    ib = tvm.tir.ir_builder.create()
+    data_ptr = ib.buffer_ptr(data)
+    argsorted_indices_ptr = ib.buffer_ptr(argsorted_indices)
+    inc_scan_ptr = ib.buffer_ptr(inc_scan)
+    unique_elements_ptr = ib.buffer_ptr(unique_elements)
+    indices_ptr = ib.buffer_ptr(indices)
+
+    batch_size = data.shape[0]
+    max_threads = tir.min(batch_size, tvm.target.Target.current(allow_none=False).max_num_threads)
+    with ib.new_scope():
+        nthread_tx = max_threads
+        nthread_bx = ceil_div(batch_size, max_threads)
+        tx = te.thread_axis("threadIdx.x")
+        bx = te.thread_axis("blockIdx.x")
+        ib.scope_attr(tx, "thread_extent", nthread_tx)
+        ib.scope_attr(bx, "thread_extent", nthread_bx)
+        tid = bx * max_threads + tx
+        with ib.if_scope(tid < batch_size):
+            indices_ptr[argsorted_indices_ptr[tid]] = inc_scan_ptr[tid]
+            with ib.if_scope(tid == 0):
+                unique_elements_ptr[inc_scan_ptr[tid]] = data_ptr[argsorted_indices_ptr[tid]]
+            with ib.else_scope():
+                with ib.if_scope(inc_scan_ptr[tid] != inc_scan_ptr[tid - 1]):
+                    unique_elements_ptr[inc_scan_ptr[tid]] = data_ptr[argsorted_indices_ptr[tid]]
+    return ib.get()
+
+
+def _calc_counts_sorted_ir(inc_scan, counts):
+    ib = tvm.tir.ir_builder.create()
+    inc_scan_ptr = ib.buffer_ptr(inc_scan)
+    counts_ptr = ib.buffer_ptr(counts)
+
+    batch_size = inc_scan.shape[0]
+    max_threads = tir.min(batch_size, tvm.target.Target.current(allow_none=False).max_num_threads)
+    with ib.new_scope():
+        nthread_tx = max_threads
+        nthread_bx = ceil_div(batch_size, max_threads)
+        tx = te.thread_axis("threadIdx.x")
+        bx = te.thread_axis("blockIdx.x")
+        ib.scope_attr(tx, "thread_extent", nthread_tx)
+        ib.scope_attr(bx, "thread_extent", nthread_bx)
+        tid = bx * max_threads + tx
+        with ib.if_scope(tid < batch_size):
+            counts_ptr[tid] = 0
+    with ib.new_scope():
+        nthread_tx = max_threads
+        nthread_bx = ceil_div(batch_size, max_threads)
+        tx = te.thread_axis("threadIdx.x")
+        bx = te.thread_axis("blockIdx.x")
+        ib.scope_attr(tx, "thread_extent", nthread_tx)
+        ib.scope_attr(bx, "thread_extent", nthread_bx)
+        tid = bx * max_threads + tx
+        atomic_add_return = ib.allocate(counts.dtype, (1,), name="atomic_add_return", scope="local")
+        with ib.if_scope(tid < batch_size):
+            index = inc_scan_ptr[tid]
+            atomic_add_return[0] = tvm.tir.call_intrin(
+                counts.dtype,
+                "tir.atomic_add",
```

Comment:
hmm I haven't looked into detail at all, but atomic sounds like a code smell... I'll think about if there is a way to avoid it

Reply:
Removed to use `arange`.

Reply:
Looks great